src/tests/end2end/SamplerTests.cpp

```diff
@@ -0,0 +1,195 @@
+// Copyright 2018 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <array>
+#include <cmath>
+
+#include "tests/NXTTest.h"
+
+#include "common/Assert.h"
+#include "common/Constants.h"
+#include "utils/NXTHelpers.h"
+
+constexpr static unsigned int kRTSize = 64;
+
+namespace {
+    struct AddressModeTestCase {
+        nxt::AddressMode mMode;
+        uint8_t mExpected2;
+        uint8_t mExpected3;
+    };
+    AddressModeTestCase addressModes[] = {
+        nxt::AddressMode::Repeat,           0, 255,
+        nxt::AddressMode::MirroredRepeat, 255,   0,
+        nxt::AddressMode::ClampToEdge,    255, 255,
+    };
+}
+
+class SamplerTest : public NXTTest {
+protected:
+
+    template<typename T>
+    nxt::Buffer MakeVertexBuffer(std::vector<T> data) {
+        return utils::CreateFrozenBufferFromData(device, data.data(), static_cast<uint32_t>(data.size() * sizeof(T)), nxt::BufferUsageBit::Vertex);
+    }
+
+    void SetUp() override {
+        NXTTest::SetUp();
+        mFB = utils::CreateBasicFramebuffer(device, kRTSize, kRTSize);
+        mFB.color.TransitionUsage(nxt::TextureUsageBit::OutputAttachment);
+
+        mBindGroupLayout = device.CreateBindGroupLayoutBuilder()
+            .SetBindingsType(nxt::ShaderStageBit::Fragment, nxt::BindingType::Sampler, 0, 1)
+            .SetBindingsType(nxt::ShaderStageBit::Fragment, nxt::BindingType::SampledTexture, 1, 1)
+            .GetResult();
+
+        auto pipelineLayout = device.CreatePipelineLayoutBuilder()
+            .SetBindGroupLayout(0, mBindGroupLayout)
+            .GetResult();
+
+        auto vsModule = utils::CreateShaderModule(device, nxt::ShaderStage::Vertex, R"(
+            #version 450
+            layout(location = 0) in vec4 pos;
+            void main() {
```

Comment:
nit: to make the test even more minimal, you could have the vertices defined in the shader, and do dead code removal. See https://github.com/Kangz/nxt-standalone/blob/master/src/tests/end2end/BlendStateTests.cpp#L36

Reply:
Done.